blog/2023-02-16-do-you-need-to-build-auth/index.md

```diff
@@ -0,0 +1,124 @@
+
+:::note 🤔 About Auth
+
+- Authentication (AuthN) answers the question “Which identity do you own?”
+- Authorization (AuthZ) answers the question “What can you do?”
+  :::
+
+I wrote an article [CIAM 101: Authentication, Identity, SSO](../2022-11-28-ciam-101-authn-identity-sso/index.mdx) to discuss these concepts in detail.
+
+### Choose authentication methods
+
+Let’s start with “authentication”, which is user sign-in in our example. Besides the username-and-password method, here are some trending methods people are adopting for a better user conversion and security:
+
+- Passwordless, i.e. dynamic verification code (via email or sms)
+- Social sign-in (Google, GitHub, etc.)
+
+The choice of method depends on the business decision, while we can take a look on the tech effort:
+
+- For passwordless, you need to find a vendor to send email or sms; then integrate with your API service (new APIs may needed).
+- For social sign-in, you must adhere to the integration guidelines of the social identity provider(s) you wish to use. Additionally, you must create a mapping between your bookstore's user IDs and the identity provider's.
+  - For example, when a user signs in with a Gmail account `foo@gmail.com`, you need to link this address to the user `foo` in the bookstore. This is to implement a unified identity system and give the user the ability to change or unlink their social identity in the future.
+
+### Design and implement sign-in experience
+
+After you decide authentication methods, a graceful and smooth sign-in experience for your end-users is the next target. The conversion here is fundamental but crucial to the business since it's a nature way to leave the persist customer data.
+
+When I worked for Airbnb, there was a whole team dedicated to optimizing the sign-in experience for multiple platforms. They conducted numerous A/B tests to determine which combination had the highest conversion rate.
+
+It’s not practical to do so when a business is getting started. But we still need to try our best to make every piece right. Which platforms would you like to run the bookstore? You may start with mobile, web, or both. The exact design will depend on the authentication methods you chose:
+
+- Username and password: The easiest one, just several input boxes and buttons.
+- Passwordless: Enter phone / email, then send and verify a dynamic code.
+- Social sign-in: Read the docs from the chosen social identity provider, follow the visual guideline, handle the redirect logic, and finally link the social identity with the bookstore identity.
+
+More things to consider to make it better:
+
+- Do you need to combine the sign-in and registration process for a specific method?
+- Do you need a “forgot password” flow to allow customers to reset their passwords independently?
+
+If you opt for native development, the workload will almost double for one additional platform.
```

Comment:
```suggestion
If you opt for native development, the workload will almost be doubled for each additional platform.
```

Reply:
double use as a verb here

Reply:
Yes, a verb, but we should use the past tense, I think

Reply:
- i think it's a natural thing to become double itself instead of "being doubled"
- "for each additional platform" may cause confusion since the base of double will be unclear

Reply:
OK